CharlieBackend.Data/Repositories/Impl/DashboardRepository.cs

```diff
@@ -0,0 +1,338 @@
+                        StudentGroupId = (long)x.Lesson.StudentGroupId,
+                        StudentId = (long)x.StudentId,
+                        LessonId = x.LessonId,
+                        LessonDate = x.Lesson.LessonDate,
+                        Presence = x.Presence,
+                    }).ToListAsync();
+
+            return studentsPresenceList;
+        }
+
+        public async Task<List<StudentMarkDto>> GetStudentsMarksListByStudentIds(IEnumerable<long> studentIds)
+        {
+            var studentsMarksList = await _applicationContext.Visits
+                    .AsNoTracking()
+                    .Where(x => x.Lesson.StudentGroup.StudentsOfStudentGroups
+                    .Any(x => studentIds.Contains(x.Student.Id)) 
+                    && x.StudentMark != null)
+                    .Select(x => new StudentMarkDto
+                    {
+                        CourseId = x.Lesson.StudentGroup.CourseId,
+                        StudentGroupId = (long)x.Lesson.StudentGroupId,
+                        StudentId = (long)x.StudentId,
+                        LessonId = x.LessonId,
+                        LessonDate = x.Lesson.LessonDate,
+                        StudentMark = x.StudentMark,
+                    }).ToListAsync();
+
+            return studentsMarksList;
+        }
+
+        public async Task<List<long>> GetGroupsIdsByStudentIdAndPeriodAsync(long studentId, 
+            DateTime? startDate, DateTime? finishDate)
+        {
+                var groupIdsbyStudentIdAndPeriod = await _applicationContext.StudentGroups
+                    .AsNoTracking()
+                    .Where(x => x.StudentsOfStudentGroups.Any(x => x.StudentId == studentId))
+                    .WhereIf(startDate != null && finishDate != default(DateTime), 
+                    x => x.StartDate >= startDate)
+                    .WhereIf(finishDate != null && finishDate != default(DateTime), 
+                    x => x.FinishDate <= finishDate)
+                    .Select(x => x.Id)
+                    .ToListAsync();
+
+                return groupIdsbyStudentIdAndPeriod;
+        }
+
+        public async Task<List<AverageStudentMarkDto>> GetStudentAverageMarksByStudentIdAsync(long studentId, List<long> studentGroupsIds)
+        {
+            var studentAverageMarskList = await _applicationContext.Visits
+                    .AsNoTracking()
+                    .Where(x => studentGroupsIds.Contains(x.Lesson.StudentGroupId.Value))
+                    .Where(x => x.StudentId == studentId && x.StudentMark != null)
+                    .Select(x => new 
+                    {
+                        CourseId = x.Lesson.StudentGroup.CourseId,
+                        StudentGroupId = (long)x.Lesson.StudentGroupId,
+                        StudentLessonMark = x.StudentMark,
+                        StudentId = x.StudentId
+                    })
+                    .AsQueryable()
```

Comment:
You don't need AsQueryable here

Reply:
But in .net 3.0 they have changed it, if i dont put AsQueryable  or toList, it gives exception https://stackoverflow.com/questions/58138556/client-side-groupby-is-not-supported